feat(common): add getState method to LocationStrategy interface

mmaterowski opened the pull request:
Add getState to LocationStrategy interface as it suppose to be the place to control all window.location interactions

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #34603


## What is the new behavior?

LocationStrategy implements getState() method which will eventually call  _window.history_ to access current history state.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
Only classes that implement Abstract `LocationStrategy` are `PathLocationStrategy` (test already present) and `HashLocationStrategy`. Currently there are no tests for `HashLocationStrategy`, if you find them necessary for this PR to merge let me know.

mmaterowski commented:
I'm having trouble understating how my change is failing following test 

> styling should set class based on priority in **styling_spec.ts**

Seems to work locally  when I execute
`KARMA_WEB_TEST_MODE=SL_REQUIRED yarn karma start ./karma-js.conf.js --single-run
`
Result:
![image](https://user-images.githubusercontent.com/23449082/163549443-d138d598-58b5-4894-9e1c-39b4a09e8806.png)

atscott reviewed (changes requested):
@mmaterowski Thanks for the PR. This does not fully address #34603 because `Location` still uses `PlatformLocation` directly rather than going through the new API you're providing here. 

In addition, you'll have to note in the commit message that this is a breaking change (add `BREAKING CHANGE: <description>` to your footer) because it adds a new required class member that any implementors of the `LocationStrategy` will need to satisfy.

mmaterowski commented on packages/common/src/location/location.ts:
```diff
@@ -62,16 +61,13 @@ export class Location implements OnDestroy {
   _urlChangeSubscription: SubscriptionLike|null = null;
 
-  constructor(platformStrategy: LocationStrategy, platformLocation: PlatformLocation) {
+  constructor(platformStrategy: LocationStrategy) {
```
Is `platformStrategy` still valid parameter name? Is there any reason no to rename it to `locationStrategy`?

atscott replied:
Yea, I think renaming it is a good idea.

atscott commented:
> I'm having trouble understating how my change is failing following test

Yea, there is known flakiness with a few tests at the moment. Sorry about the friction.

mmaterowski committed:
feat(common): add getState method to LocationStrategy interface

Adds getState to LocationStrategy interface as it suppose to be the place to control all window.location interactions.

BREAKING CHANGE:
Adds new required class member that any implementors of the LocationStrategy will need to satisfy.
Location does not depend on PlatformLocation anymore.

mmaterowski commented:
@atscott thanks for reviewing my PR.
Since `PlatformLocation` is no longer used in `Location` I removed it from Location's constructor. 
Breaking changes footer added.

atscott reviewed (approved):
LGTM 👍

jessicajaniuk reviewed (approved):
reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

atscott commented:
Note for team members: Since this is a breaking change, I've triggered [global presubmits](http://tap/442050757) to identify what needs to be updated. 

Also adding the blocked label until they necessary internal cleanup has been completed.

dylhunn reviewed (approved):
reviewed-for: public-api

dylhunn commented:
@atscott by internal cleanup, do you mean TGP breakage? The linked presubmit is not showing any real failures.

atscott commented:
@dylhunn The TGP isn't finished (the full train runs don't show up in the UI until they are; only the smoke test is available immediately). From prior work in this area, I know there will be failures that need to be fixed.

dylhunn commented:
> only the smoke test is available immediately

Thanks for clarifying!

atscott commented:
Internal cleanup is finished. Started a new global presubmit to ensure there weren't any regressions. Then this should be good to go today or tomorrow.

atscott commented:
merge assistance: requires internal patch when syncing: http://cl/443689787

atscott commented:
This PR was merged into the repository by commit 31d7c3bd71fdab3fef1b4615ecb8124fe6c165bd.
